lib/navigation/preprocess_imu.cpp

```diff
@@ -1,56 +1,123 @@
 #include "preprocess_imu.hpp"
 
+#include <cmath>
+
+#include <algorithm>
+#include <numeric>
+#include <optional>
+
 namespace hyped::navigation {
 
 ImuPreprocessor::ImuPreprocessor()
 {
   // TODOLater: implement
 }
 
-core::ImuData ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
+std::optional<core::ImuData> ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkReliable
-
-  */
-  return {0, 0, 0, 0};
+  core::ImuData imu_data;
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(raw_imu_data.at(i).at(j), 2);
+    }
+    imu_data.at(i) = std::sqrt(magnitude);
+  }
+
+  const core::ImuData accelerometer_data = detectOutliers(imu_data);
+  checkReliable();
+  return accelerometer_data;
 }
 
-core::ImuData ImuPreprocessor::detectOutliers(const core::RawImuData imu_data)
+core::ImuData ImuPreprocessor::detectOutliers(const core::ImuData imu_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  outlier_imus_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::ImuData accelerometer_data;
+  for (std::size_t i; i < core::kNumImus; ++i) {
+    accelerometer_data.at(i) = imu_data.at(i);
+  }
+
+  Quartiles quartiles;
+  if (num_reliable_accelerometers_ == core::kNumImus) {
+    quartiles = getQuartiles(imu_data);
+  } else {
+    std::array<core::Float, core::kNumImus - 1> faulty_data;
+    std::size_t index;
+    for (std::size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (!are_imus_reliable_.at(i)) { index = i; }
+    }
+    for (std::size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (i >= index) {
+        faulty_data.at(i) = imu_data.at(i + 1);
+      } else {
+        faulty_data.at(i) = imu_data.at(i);
+      }
+    }
+    quartiles = getQuartiles(faulty_data);
+  }
+
+  const core::Float iqr         = quartiles.q3 - quartiles.q1;
```

Comment:
move these assignments into if-else statements, changing threshold to 1.2 if an imu is faulty